lib/ukstore/include/uk/store.h

```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+typedef void (*_uk_store_entry_freed_func_t)(void *);
 
 /* Flags if an entry is static or dynamic */
 #define UK_STORE_ENTRY_FLAG_STATIC	(1 << 0)
 
 #if !CONFIG_LIBUKSTORE
 
 /* Do not call directly */
-#define __UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set,	\
-				arg_cookie)				\
+#define __UK_STORE_STATIC_ENTRY(eid, entry, lib_str, e_type, e_get, e_set)\
 	static const struct uk_store_entry				\
 	__unused __uk_store_entries_list ## _ ## entry = {		\
-		.name = STRINGIFY(entry),				\
-		.type       = (UK_STORE_ENTRY_TYPE(e_type)),		\
-		.get.e_type = (e_get),					\
-		.set.e_type = (e_set),					\
-		.flags      = UK_STORE_ENTRY_FLAG_STATIC,		\
-		.cookie     = (arg_cookie)				\
+		.id             = eid,					\
+		.name           = STRINGIFY(entry),			\
+		.type           = (UK_STORE_ENTRY_TYPE(e_type)),	\
+		.ops.get.e_type = (e_get),				\
+		.ops.set.e_type = (e_set),				\
+		.flags          = UK_STORE_ENTRY_FLAG_STATIC		\
 	}
 
 /* Do not call directly */
-#define _UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set,	\
-				cookie)					\
-	__UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set, cookie)
+#define _UK_STORE_STATIC_ENTRY(id, entry, lib_str, e_type, e_get, e_set)\
+	__UK_STORE_STATIC_ENTRY(id, entry, lib_str, e_type, e_get, e_set)
 
 /**
  * Adds an entry to the entry section of a library.
  *
+ * @param id entry id
  * @param entry the entry in the section
  * @param e_type type for the entry, e.g., s8, u16, charp
  * @param e_get getter pointer (optional, can be NULL)
  * @param e_set setter pointer (optional, can be NULL)
- * @param cookie a cookie for extra storage
  */
-#define UK_STORE_STATIC_ENTRY(entry, e_type, e_get, e_set, cookie)	\
-	_UK_STORE_STATIC_ENTRY(entry, STRINGIFY(__LIBNAME__),		\
-		e_type, e_get, e_set, cookie)
+#define UK_STORE_STATIC_ENTRY(id, entry, e_type, e_get, e_set)		\
+	_UK_STORE_STATIC_ENTRY(id, entry, STRINGIFY(__LIBNAME__),	\
+			       e_type, e_get, e_set)
 
 #else /* !CONFIG_LIBUKSTORE */
 
 /* Library-specific information generated at compile-time */
 #include <uk/bits/store_libs.h>
 
+struct uk_store_ops {
+	/* Function getter pointer */
+	union {
+		_uk_store_get_s8_func_t     s8;
+		_uk_store_get_u8_func_t     u8;
+		_uk_store_get_s16_func_t    s16;
+		_uk_store_get_u16_func_t    u16;
+		_uk_store_get_s32_func_t    s32;
+		_uk_store_get_u32_func_t    u32;
+		_uk_store_get_s64_func_t    s64;
+		_uk_store_get_u64_func_t    u64;
+		_uk_store_get_uptr_func_t   uptr;
+		_uk_store_get_charp_func_t  charp;
+	} get;
+
+	/* Function setter pointer */
+	union {
+		_uk_store_set_s8_func_t    s8;
+		_uk_store_set_u8_func_t    u8;
+		_uk_store_set_s16_func_t   s16;
+		_uk_store_set_u16_func_t   u16;
+		_uk_store_set_s32_func_t   s32;
+		_uk_store_set_u32_func_t   u32;
+		_uk_store_set_s64_func_t   s64;
+		_uk_store_set_u64_func_t   u64;
+		_uk_store_set_uptr_func_t  uptr;
+		_uk_store_set_charp_func_t charp;
+	} set;
+
+	_uk_store_entry_freed_func_t cleanup;
```

Comment:
Remove?

Reply:
We don't use it so far, will remove for now

Reply:
:+1: